# 定义通用返回



## 定义封装类型Result

```java
package cn.xinyuan.blog.common.api;

import cn.xinyuan.blog.common.exception.ErrorEnum;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

/**
 * @ClassName: Result
 * @Description: 通用返回对象
 * @Author: xinyuan
 * @CreateDate: 2020/1/22 15:35
 */
@Data
@AllArgsConstructor
@NoArgsConstructor
public class Result<T> {
    private int code;
    private String message;
    private T data;

    /**
     * 成功返回结果
     *
     * @param data 获取的数据
     */
    public static <T> Result<T> success(T data) {
        return new Result<T>(ErrorEnum.SUCCESS.getCode(), ErrorEnum.SUCCESS.getMsg(), data);
    }

    /**
     * 成功返回结果
     *
     * @param data 获取的数据
     * @param  message 提示信息
     */
    public static <T> Result<T> success(T data, String message) {
        return new Result<T>(ErrorEnum.SUCCESS.getCode(), message, data);
    }

    /**
     * 失败返回结果
     * @param errorCode 错误码
     */
    public static <T> Result<T> failed(ErrorEnum errorCode) {
        return new Result<T>(errorCode.getCode(), errorCode.getMsg(), null);
    }

    /**
     * 失败返回结果
     * @param message 提示信息
     */
    public static <T> Result<T> failed(String message) {
        return new Result<T>(ErrorEnum.FAILED.getCode(), message, null);
    }

    /**
     * 失败返回结果
     */
    public static <T> Result<T> failed() {
        return failed(ErrorEnum.FAILED.getMsg());
    }



}
```



## 状态码枚举类

ErrorEnum

```java
package cn.xinyuan.blog.common.exception;

import lombok.AllArgsConstructor;
import lombok.Getter;

/**
 * @ClassName: ErrorEnum
 * @Description: java类作用描述
 * @Author: xinyuan
 * @CreateDate: 2020/1/22 15:31
 */
@Getter
@AllArgsConstructor
public enum ErrorEnum {
    // 系统错误
    SUCCESS(200, "操作成功"),
    FAILED(500,"操作失败"),
    PATH_NOT_FOUND(404,"路径不存在，请检查路径"),
    NO_AUTH(403,"没有权限，请联系管理员"),
    DUPLICATE_KEY(501,"数据库中已存在该记录"),
    TOKEN_GENERATOR_ERROR(502,"token生成失败"),
    NO_UUID(503,"uuid为空"),
    SQL_ILLEGAL(504,"sql非法"),

    //用户权限错误
    INVALID_TOKEN(1001,"token不合法"),

    //登录模块错误
    LOGIN_FAIL(10001,"登录失败"),
    CAPTCHA_WRONG(10002,"验证码错误"),
    USERNAME_OR_PASSWORD_WRONG(10003,"用户名或密码错误"),
    ACCOUNT_LOCK(10004, "账号已被锁定，请联系管理员"),

    //七牛OSS错误
    OSS_CONFIG_ERROR(10050,"七牛配置信息错误"),
    OSS_UPLOAD_ERROR(10051,"OSSBookNote上传失败");

    private int code;
    private String msg;
}
```

## Result使用示例

返回内容

```java
return Result.success("success");
```

返回的状态码信息

```java
return Result.failed(ErrorEnum.USERNAME_OR_PASSWORD_WRONG);
```

# 定义统一异常处理

## 全局异常处理

定义一个处理类，使用@ControllerAdvice注解。

@ControllerAdvice注解：控制器增强，一个被@Component注册的组件。

配合@ExceptionHandler来增强所有的@requestMapping方法。

例如：@ExceptionHandler(Exception.class)  用来捕获@requestMapping的方法中所有抛出的exception。

代码：

```java
@ControllerAdvice
public class GlobalDefultExceptionHandler {
    //声明要捕获的异常
    @ExceptionHandler(Exception.class)
    @ResponseBody
    public String defultExcepitonHandler(HttpServletRequest request,Exception e) {
        return “error”;
    }
}
```

这样，全局异常处理类完毕。可以添加自己的逻辑。

实际项目使用：

```java
package cn.xinyuan.blog.common.exception;

import cn.xinyuan.blog.common.api.Result;
import lombok.extern.slf4j.Slf4j;
import org.apache.shiro.authz.AuthorizationException;
import org.springframework.dao.DuplicateKeyException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.servlet.NoHandlerFoundException;

/**
 * MyExceptionHandler
 *
 * @author xinyuan
 * @description 统一异常处理器
 */
@RestControllerAdvice
@Slf4j
public class MyExceptionHandler {

    /**
     * 处理自定义异常
     * @param e
     * @return
     */
    @ExceptionHandler(MyException.class)
    public Result handleMyException(MyException e){
        Result result=new Result();
        result.setCode(e.getCode());
        result.setMessage(e.getMsg());
        return result;
    }

    @ExceptionHandler(NoHandlerFoundException.class)
    public Result handlerNoFoundException(Exception e){
        log.error(e.getMessage(),e);
        return Result.failed(ErrorEnum.PATH_NOT_FOUND);
    }

    @ExceptionHandler(DuplicateKeyException.class)
    public Result handleDuplicateKeyException(DuplicateKeyException e){
        log.error(e.getMessage(),e);
        return Result.failed(ErrorEnum.DUPLICATE_KEY);
    }

    @ExceptionHandler(AuthorizationException.class)
    public Result handleAuthorizationException(AuthorizationException e){
        log.error(e.getMessage(),e);
        return Result.failed(ErrorEnum.NO_AUTH);
    }

    @ExceptionHandler(Exception.class)
    public Result handleException(Exception e){
        log.error(e.getMessage(),e);
        return Result.failed();
    }
}

```

解析：

- 处理自定义异常

MyException是自定义的异常

```java
    /**
     * 处理自定义异常
     * @param e
     * @return
     */
    @ExceptionHandler(MyException.class)
    public Result handleMyException(MyException e){
      	Result result=new Result();
        result.setCode(e.getCode());
        result.setMessage(e.getMsg());
        return result;
    }
```

上面节选的这段代码就是处理自定义的异常类，截取到了就以这种方式返回。

## 自定义业务异常类

还有一个问题，有的时候,我们需要业务逻辑时抛出自定义异常，这个时候需要自定义业务异常类。

定义class：BusinessException ,使他继承于RuntimeException.

说明：因为某些业务需要进行业务回滚。但spring的事务只针对RuntimeException的进行回滚操作。所以需要回滚就要继承RuntimeException。

```java
public class BusinessException extends RuntimeException{

}
```

实际项目使用：

```java
package cn.xinyuan.blog.common.exception;

import lombok.Data;

/**
 * MyException
 *
 * @author xinyuan
 * @description 自定义异常
 */
@Data
public class MyException extends RuntimeException{
    private String msg;
    private int code = 500;

    public MyException(){
        super(ErrorEnum.FAILED.getMsg());
        this.msg=ErrorEnum.FAILED.getMsg();
    }

    public MyException(ErrorEnum eEnum,Throwable e){
        super(eEnum.getMsg(),e);
        this.msg=eEnum.getMsg();
        this.code=eEnum.getCode();
    }

    public MyException(ErrorEnum eEnum){
        this.msg=eEnum.getMsg();
        this.code=eEnum.getCode();
    }

    public MyException(String exception){
       this.msg=exception;
    }
}
```



